tools/schemacode/bidsschematools/validator.py

```diff
@@ -1,31 +1,147 @@
 import bidsschematools.utils
 
 lgr = bst.utils.get_logger()
 
-# The list of which entities create directories could be dynamically specified by the YAML, but for
-# now, it is not.
-# Ordering is important, as "subject" follows "session" alphabetically, but is hierarchically
-# above it.
-DIR_ENTITIES = ["subject", "session"]
+BUNDLED_SCHEMA_PATH = "{module_path}/data/schema"
+VALIDATOR_SCHEMA_COMPATIBILITY_LEVEL = "minor"
+
+
+def _bids_schema_versioncheck(schema_dir, compatibility=VALIDATOR_SCHEMA_COMPATIBILITY_LEVEL):
+    """
+    Use the `SCHEMA_VERSION` descriptor file to determine whether a selected schema directory is
+    compatible with the validator.
+
+    Parameters
+    ----------
+    schema_dir : str
+        A string which specifies a path.
+    compatibility : "major" or "minor", optional
+        Compatibility range of the validator. Major means "M.*" are considered compatible, minor
+        means "M.m.*" are considered compatible.
+
+    Returns
+    -------
+    bool:
+        Whether the schema is compatible with the validator.
+    """
+    schema_version_file = os.path.join(schema_dir, "SCHEMA_VERSION")
+    try:
+        with open(schema_version_file, "r") as f:
+            schema_version = f.readlines()[0].strip()
+            if _bids_schema_versioncheck(schema_version_file):
+                if compatibility == "major":
+                    if schema_version.rsplit(".", 1)[0] == bst.__version__.rsplit(".", 1)[0]:
+                        return True
+                elif compatibility == "minor":
+                    if schema_version.rsplit(".", 2)[:2] == bst.__version__.rsplit(".", 2)[:2]:
+                        return True
+                else:
+                    raise ValueError(
+                        "Schema compatibility needs to be set to either “major” or “minor”."
+                    )
+                lgr.warning(
+                    "The selected schema `%s`, has a schema version (`%s`) which is "
+                    "incompatible with the validator. Attempting to query the BIDS reference "
+                    "for compatible versions.",
+                    schema_dir,
+                )
+    except FileNotFoundError:
+        lgr.warning(
+            "The selected schema directory, `%s`, does not contain a SCHEMA_VERSION file. "
+            "Cannot ascertain compatibility. Attempting to query the BIDS reference for "
+            "compatible versions.",
+            schema_dir,
+        )
+    return False
+
+
+def _find_bids_root(in_paths, accept_non_bids_dir):
+    """
+    Return BIDS root for a list of paths.
+    Raise error if more than one root is found and, optionally, if none is found.
+    """
+
+    dataset_descriptions = []
+    for in_path in in_paths:
+        in_path = os.path.abspath(os.path.expanduser(in_path))
+        dataset_description = _find_dataset_description(in_path)
+        if dataset_description and dataset_description not in dataset_descriptions:
```

Comment:
solved in [d632300](https://github.com/bids-standard/bids-specification/pull/1398/commits/d632300b5ecde86aec0b443217d71ff93796247d).

Reply:
Well, no, actually. Using a set makes it more difficult to extract the value at the end, and we can't just assert the length since we want to display different logging output depending on whether it's length-0 or length->1. I think just leaving it as a list is preferable.